components/Components/URLSearchParams.php

```diff
@@ -0,0 +1,428 @@
+<?php
+
+/**
+ * League.Uri (https://uri.thephpleague.com)
+ *
+ * (c) Ignace Nyamagana Butera <nyamsprod@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace League\Uri\Components;
+
+use ArgumentCountError;
+use Closure;
+use Countable;
+use Iterator;
+use IteratorAggregate;
+use League\Uri\Contracts\QueryInterface;
+use League\Uri\Contracts\UriComponentInterface;
+use League\Uri\Contracts\UriInterface;
+use League\Uri\KeyValuePair\Converter;
+use League\Uri\QueryString;
+use League\Uri\Uri;
+use Psr\Http\Message\UriInterface as Psr7UriInterface;
+use Stringable;
+use TypeError;
+
+use function array_map;
+use function array_reduce;
+use function count;
+use function func_get_arg;
+use function func_num_args;
+use function is_iterable;
+use function is_string;
+use function str_starts_with;
+
+/**
+ * @see https://url.spec.whatwg.org/#interface-urlsearchparams
+ *
+ * @implements IteratorAggregate<array{0:string, 1:string}>
+ */
+final class URLSearchParams implements Countable, IteratorAggregate, UriComponentInterface
+{
+    private QueryInterface $query;
+
+    /**
+     * New instance.
+     *
+     * A string, which will be parsed from application/x-www-form-urlencoded format. A leading '?' character is ignored.
+     * A literal sequence of name-value string pairs, or any object with an iterator that produces a sequence of string pairs.
+     * A record of string keys and string values. Note that nesting is not supported.
+     */
+    public function __construct(object|array|string|null $query = '')
```

Comment:
Should the type really be `object` here or be more specific like `Stringable|\Traversable|\QueryInterface|...`?

Reply:
Yes it should because according to the specification any object that can be iterable should be acceptable which translate to me in PHP by traversable , stdClass or any object with public properties as stated by the example in the specification and the fact that it works for instance for DateInterval. Hence why I added more named constructors in the PHP implementation to restrict in code what can and can not be use

Reply:
Example from MDN `URLSearchParams`:

```js
// Retrieve params via url.search, passed into constructor
const url = new URL("https://example.com?foo=1&bar=2");
const params1 = new URLSearchParams(url.search);

// Get the URLSearchParams object directly from a URL object
const params1a = url.searchParams;

// Pass in a string literal
const params2 = new URLSearchParams("foo=1&bar=2");
const params2a = new URLSearchParams("?foo=1&bar=2");

// Pass in a sequence of pairs
const params3 = new URLSearchParams([
  ["foo", "1"],
  ["bar", "2"],
]);

// Pass in a record
const params4 = new URLSearchParams({ foo: "1", bar: "2" });
```

Which I would translate in PHP as the following:

```php
// Retrieve params via url.search, passed into constructor
$params1 = URLSearchParams::fromUri("https://example.com?foo=1&bar=2");

// Pass in a string literal
$params2 = new URLSearchParams("foo=1&bar=2");
$params2a = new URLSearchParams("?foo=1&bar=2");

// Pass in a sequence of pairs
$params3 = new URLSearchParams([
  ["foo", "1"],
  ["bar", "2"],
]);

// Pass in a record
$params4 = new URLSearchParams((object)["foo" => "1", "bar" => "2" });
//which means
$params5 = new URLSearchParams(new DateInterval('P3MT12M5S')); is also possible.
```

Hence why to reduce subtle errors I've added

- `URLSearchParams::new` for string query with or without delimiter.
- `URLSearchParams::fromUri` for uri string or object.
- `URLSearchParams::fromPairs` for iterable of pairs elements.
- `URLSearchParams::fromRecords` for records. I translate records as objects with public properties or generic iterator of key/value.
- The `URLSearchParams::fromParameters` for array produces by `parse_str` and/or supported by `http_build_query`